Allow setting a storage limit for all origins

szewai opened the pull request:
#### e9a1f73b15efbe97e1d3a5ea7a5d211c616215a9
<pre>
Allow setting a storage limit for all origins
<a href="https://bugs.webkit.org/show_bug.cgi?id=254011">https://bugs.webkit.org/show_bug.cgi?id=254011</a>
rdar://106792298

Reviewed by Youenn Fablet.

Make it possible to set a total quota of all origins based on disk space. This is implemented by adding a setting
totalQuotaRatio to WebsiteDataStoreConfiguration. The overall quota of a data store is calculated by:
totalDiskSpace * totalQuotaRatio. When this quota is exceeded, network process will evict data (only for data types that
are managed by NetworkStorageManager) based on LRU policy.

Test: WKWebsiteDataStoreConfiguration.TotalQuotaRatio

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp: Added.
(WebKit::NetworkQuotaManager::NetworkQuotaManager):
(WebKit::NetworkQuotaManager::originUsageUpdated):
(WebKit::NetworkQuotaManager::performEvictionIfNeeded):
(WebKit::NetworkQuotaManager::originVisited):
* Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.h: Copied from Source/WebKit/Platform/IPC/WorkQueueMessageReceiver.h.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::canHandleTypes):
(WebKit::NetworkStorageManager::allManagedTypes):
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::quotaManager):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::removeOriginStorageManagerIfPossible):
(WebKit::NetworkStorageManager::updateOriginModificationTime):
(WebKit::NetworkStorageManager::evictDataByTopOrigin):
(WebKit::NetworkStorageManager::getOriginDirectory):
(WebKit::NetworkStorageManager::requestSpace):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp: Renamed from Source/WebKit/NetworkProcess/storage/QuotaManager.cpp.
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::usage):
(WebKit::OriginQuotaManager::requestSpace):
(WebKit::OriginQuotaManager::handleRequests):
(WebKit::OriginQuotaManager::grantWithCurrentQuota):
(WebKit::OriginQuotaManager::usageUpdated):
(WebKit::OriginQuotaManager::grantFastPath):
(WebKit::OriginQuotaManager::didIncreaseQuota):
(WebKit::OriginQuotaManager::resetQuotaUpdatedBasedOnUsageForTesting):
(WebKit::OriginQuotaManager::resetQuotaForTesting):
(WebKit::OriginQuotaManager::reportedQuota const):
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h: Renamed from Source/WebKit/NetworkProcess/storage/QuotaManager.h.
(WebKit::OriginQuotaManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::isActive const):
(WebKit::OriginStorageManager::StorageBucket::hasDataInMemory const):
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::quotaManager):
(WebKit::OriginStorageManager::fileSystemStorageManager):
(WebKit::OriginStorageManager::idbStorageManager):
(WebKit::OriginStorageManager::cacheStorageManager):
(WebKit::OriginStorageManager::backgroundFetchManager):
(WebKit::OriginStorageManager::hasDataInMemory):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/Platform/IPC/WorkQueueMessageReceiver.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration totalQuotaRatio]):
(-[_WKWebsiteDataStoreConfiguration setTotalQuotaRatio:]):
(-[_WKWebsiteDataStoreConfiguration volumeCapacityOverride]):
(-[_WKWebsiteDataStoreConfiguration setVolumeCapacityOverride:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::totalQuotaRatio const):
(WebKit::WebsiteDataStoreConfiguration::setTotalQuotaRatio):
(WebKit::WebsiteDataStoreConfiguration::volumeCapacityOverride const):
(WebKit::WebsiteDataStoreConfiguration::setVolumeCapacityOverride):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/262196@main">https://commits.webkit.org/262196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7d2e955d48cb24e1577a4139bf4fc13ff6840ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/886 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/910 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1257 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/783 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/874 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/995 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/993 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/832 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/829 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1194 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/824 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/828 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/802 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/850 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1852 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/843 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/789 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/835 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/203 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/856 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,81 @@
+{
+    for (auto&& [origin, usage] : usages) {
+        m_origins.add(origin);
+        m_originUsageMap.set(origin, usage);
```
Given usages is an r-value, we could move origin here.
Or we could use a const Vector&

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,81 @@
+void NetworkQuotaManager::originUsageUpdated(const WebCore::ClientOrigin& origin, uint64_t usage)
+{
+    auto iter = m_originUsageMap.find(origin);
+    if (iter != m_originUsageMap.end() && iter->value == usage)
```
Is there a case for ASSERT(m_originUsageMap.contains(origin))?

szewai replied:
I don't think we should `ASSERT(m_originUsageMap.contains(origin)).`
If `!m_originUsageMap.contains(origin)`, it means it is a new origin (at least from NetworkQuotaManager's point of view), which is a valid case. With current implementation (that originUsageUpdated is called when QuotaManager updates its usage member, its possible usage is zero, e.g. QuotaManager initializes its usage).

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -290,6 +299,45 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        Vector<std::pair<WebCore::ClientOrigin, WallTime>> origins;
+        forEachOriginDirectory([&](auto directory) {
+            auto originFile = originFilePath(directory);
+            auto modificationTime = FileSystem::fileModificationTime(originFile);
```
How are we updating the modification time of the origin file?

szewai replied:
Every time `OriginStorageManager` is created, it will try overwriting origin file, see `NetworkStorageManager::writeOriginToFileIfNecessary`.

szewai replied:
The latest patch updates origins list when `writeOriginToFileIfNecessary()` is invoked, i.e. when there's a storage operation that ensures OriginStorageManager

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
```diff
@@ -750,4 +750,84 @@ function sendMessage(message) { \
+    receivedScriptMessage = false;
+    EXPECT_WK_STREQ(@"error", [lastScriptMessage body]);
+
+    // Ensure webkit.org data is evicted.
```
Can we test 3 origins, something like?
- load origin1
- load origin2
- reload origin1 to update LRU
- load origin 3 and verify origin2 is evicted.

And maybe the same test but with origin 2 page being opened so that origin 1 gets evicted?

szewai replied:
I can try: one thing a bit tricky is the quota is set based on disk space, so the test might require different values on different space. I guess I could introduce a mock disk space value.

szewai replied:
Updated test with 3 origins.

youennf commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
```diff
@@ -83,6 +83,7 @@ class OriginStorageManager : public CanMakeWeakPtr<OriginStorageManager> {
     void closeCacheStorageManager();
     String resolvedPath(WebsiteDataType);
     bool isActive();
+    bool hasDataInMemory();
```
const, probably for isActive and isEmpty as well.

szewai replied:
They are not const because they might initialize `m_defaultBucket`.

youennf commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
```diff
@@ -85,6 +85,7 @@ class OriginStorageManager::StorageBucket {
     void closeCacheStorageManager();
     bool isActive() const;
+    bool hasDataInMemory() const;
     bool isEmpty();
```
const?

szewai replied:
Not quite related but sure

szewai replied:
hmm actually this function requires directory to be resolved, so cannot be const.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,81 @@
+    if (iter != m_originUsageMap.end() && iter->value == usage)
+        return;
+
+    m_origins.appendOrMoveToLast(origin);
```
I could see a service worker doing a lot of write operations at init time but would then only read data for a long time. Can we compute LRU based on any data operation, read or write instead of just write operations?
Or maybe origin navigation based LRU would be simpler and good enough?

szewai replied:
I've updated the patch to make origin list updated on operations that requires OriginStorageManager to be present.
We could improve later if needed.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.h:
```diff
@@ -0,0 +1,55 @@
+
+class NetworkStorageManager;
+
+class NetworkQuotaManager : public ThreadSafeRefCounted<NetworkQuotaManager> {
```
This class does not seem thread safe, can we keep it as a simple RefCounted?

szewai replied:
It seems unnecessary to make it RefCounted; I've dropped it.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.h:
```diff
@@ -0,0 +1,55 @@
+    void originUsageUpdated(const WebCore::ClientOrigin&, uint64_t usage);
+    
+private:
+    NetworkQuotaManager(uint64_t quota, Vector<std::pair<WebCore::ClientOrigin, uint64_t>>&&, Function<bool(const WebCore::ClientOrigin&)>&&);
```
We have NetworkQuotaManager and QuotaManager.
I guess we could rename QuotaManager as OriginQuotaManager or something like that.

szewai replied:
Renamed QuotaManager to OriginQuotaManager.

youennf commented on Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
```diff
@@ -507,6 +517,8 @@ std::optional<NetworkSessionCreationParameters> NetworkSessionCreationParameters
         , WTFMove(*perThirdPartyOriginStorageQuota)
         , WTFMove(*originQuotaRatio)
+        , WTFMove(*totalQuotaRatio)
+        , WTFMove(*mockVolumeCapacity)
```
Maybe we should rename to volumeCapacityOverride?

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -288,6 +298,50 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        }
+
+        uint64_t quota = std::numeric_limits<uint64_t>::max();
+        if (m_totalQuotaRatio) {
```
Shouldn't we use 1 if m_totalQuotaRatio is not set?

szewai replied:
We might fail to read volume capacity (file system error), so I pick the max value as default.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -288,6 +298,50 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        m_quotaManager = makeUnique<NetworkQuotaManager>(quota, usages, [weakThis = WeakPtr { *this }](auto origin) {
+            if (!weakThis)
+                return false;
+            return weakThis->evictData(origin);
```
One liner, `return weakThis ? ...`

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -266,6 +275,7 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
 #if PLATFORM(IOS_FAMILY)
         includeOriginInBackupIfNecessary(*manager);
 #endif
+        updateOriginModificationTime(origin);
```
So we update the origin file everytime we create a storage manager.
This means that when we restart the networking process, that is that order that we will use, which is potentially different from the past origin storage manager in memory list?
This is probably fine as eviction is hopefully an edge case to hit the quota and storing the list would be overkill.

I wonder whether we could not piggy back on what ITP is doing here, since it is keeping the last time an origin was visited, to clear them after 7 days. Do you know where this computation happen and if it would be easy to reuse it?

szewai replied:
> So we update the origin file everytime we create a storage manager.

It is every time` NetworkStorageManager::originStorageManager(const WebCore::ClientOrigin& origin, ShouldWriteOriginFile shouldWriteOriginFile)` is called (OriginStorageManager may already exist).

I think ITP data is in `ResourceLoadStatisticsDatabaseStore`. It seems to not record when an origin is visited, but when an origin has user interaction (See `DomainData` in ResourceLoadStatisticsDatabaseStore.h).
I thought about this, a problem is ITP is only available on Cocoa platforms and only works when tracking prevention is enabled (e.g. in Safari users can turn off cross-site tracking prevention.)

I think we could use ITP but we probably still need a fallback implementation when it's not available.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,108 @@
+    performQuotaCheck();
+}
+
+void NetworkQuotaManager::performQuotaCheck()
```
We are not really doing a quota check.
How about validateTotalUsage, or ensureTotalUsageIsBelowQuota?

szewai replied:
Changed to performEvictionIfNeeded

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,108 @@
+        HashMap<WebCore::SecurityOriginData, uint64_t> failedClientOrigins;
+        bool failed = false;
+        for (auto& [origin, usage] : originUsage) {
+            auto temporigin = WebCore::ClientOrigin { topOrigin, origin };
```
s/temporigin/tempOrigin/

szewai replied:
This is unused; removed.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,108 @@
+        bool failed = false;
+        for (auto& [origin, usage] : originUsage) {
+            auto temporigin = WebCore::ClientOrigin { topOrigin, origin };
+            if (m_evictDataFunction(WebCore::ClientOrigin { topOrigin, origin }))
```
Could we simplify this by having evictDataFunction only taking the top origin and doing the work to find all the relevant full origins?
This class is somehow recreating a map of origins and we could keep doing that in a single place NetworkStorageManager. Or we could just take benefit of the fact that `top origin` is the root folder and contain all client origins whose topOrigin is `top origin`.

Is there a use case for clearing (A, B) but not (A, C) for instance?

szewai replied:
With current patch, for top-level origin A to be evicted, if origin (A,B) is active (e.g. has CacheStorage connection), but (A, C) is not, we will delete (A, B)'s data but not (A, C)'s.
It probably makes sense to keep/delete them at the same time, since we update the LRU list when any of them is updated. 

I will make the change to let m_evictDataFunction take top origin.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -300,19 +354,26 @@ OriginStorageManager& NetworkStorageManager::originStorageManager(const WebCore:
+                increaseQuotaFunction = { };
+            } else if (auto capacity = FileSystem::volumeCapacity(m_path)) {
                 quota = m_originQuotaRatio.value() * capacity.value();
                 increaseQuotaFunction = { };
```
We are doing this computation here and above, can we do it in one single place?
Maybe we should have a m_totalQuota?

szewai replied:
This is for origin quota. We only compute total quota once when creating `NetworkQuotaManager`.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,108 @@
+    auto currentUsage = originUsage.ensure(origin.clientOrigin, [] {
+        return 0;
+    }).iterator->value;
+    originUsage.set(origin.clientOrigin, usage);
```
If you have the iterator, it would be good to use it to set usage, instead of using set just after.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp:
```diff
@@ -0,0 +1,108 @@
+        m_origins.appendOrMoveToLast(origin.topOrigin);
+        m_originUsageMap.ensure(origin.topOrigin, [] {
+            return HashMap<WebCore::SecurityOriginData, uint64_t> { };
+        }).iterator->value.set(origin.clientOrigin, usage);
```
s/set/add/

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -288,6 +298,50 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+            auto originFile = originFilePath(directory);
+            auto modificationTime = FileSystem::fileModificationTime(originFile);
+            if (modificationTime) {
+                if (auto origin = readOriginFromFile(originFile))
```
Should we delete folders for which we are not able to read origin?

szewai replied:
It's a bit tricky. If an origin file is not written, but the origin folder has data, there's a chance that next time when the origin is visited, the origin file can be created correctly and the data can be used by that origin.

If you think we should remove it, I can file a radar and follow up to remove the folder when origin file does not exist.

youennf replied:
If the origin file is not readable, we might want to remove the data and will not do it.
Later, the data might be `resurrected` if the salt is the same.
We should at least file a bug to think of the best way to do things.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -340,6 +401,24 @@ bool NetworkStorageManager::removeOriginStorageManagerIfPossible(const WebCore::
+{
+    assertIsCurrent(workQueue());
+    // This function should be called when origin is in use, i.e. OriginStorageManager exists.
+    auto* manager = m_originStorageManagers.get(origin);
```
Am I correct that it is only called for each write operations, while it is only called for the first read operation during the lifetime of the network process?
This makes sense from a perf perspective, but tends to favour applications writing often compared to applications writing once and reading often.

szewai replied:
Not really, this is called in `writeOriginToFileIfNecessary()`, and `writeOriginToFileIfNecessary()` is called in `originStorageManager()`, and `originStorageManager()` can be called when there is storage client needs to access `OriginStorageManager` (you may search `originStorageManager()` in `NetworkStorageManager.cpp` if you want to know when it could be called).

youennf reviewed (approved):
OK, let's go with this approach.
I still think we should try to make it clear what our LRU policy is, how we are computing it.
Currently, it can change very easily without us noticing it.

Also, I would think we should remove the whole top origin and all its children.
If a third party origin is active, the top origin should somehow be active as well.

szewai committed:
Allow setting a storage limit for all origins
https://bugs.webkit.org/show_bug.cgi?id=254011
rdar://106792298

Reviewed by Youenn Fablet.

Make it possible to set a total quota of all origins based on disk space. This is implemented by adding a setting
totalQuotaRatio to WebsiteDataStoreConfiguration. The overall quota of a data store is calculated by:
totalDiskSpace * totalQuotaRatio. When this quota is exceeded, network process will evict data (only for data types that
are managed by NetworkStorageManager) based on LRU policy.

Test: WKWebsiteDataStoreConfiguration.TotalQuotaRatio

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.cpp: Added.
(WebKit::NetworkQuotaManager::NetworkQuotaManager):
(WebKit::NetworkQuotaManager::originUsageUpdated):
(WebKit::NetworkQuotaManager::performEvictionIfNeeded):
(WebKit::NetworkQuotaManager::originVisited):
* Source/WebKit/NetworkProcess/storage/NetworkQuotaManager.h: Copied from Source/WebKit/Platform/IPC/WorkQueueMessageReceiver.h.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::canHandleTypes):
(WebKit::NetworkStorageManager::allManagedTypes):
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::quotaManager):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::removeOriginStorageManagerIfPossible):
(WebKit::NetworkStorageManager::updateOriginModificationTime):
(WebKit::NetworkStorageManager::evictDataByTopOrigin):
(WebKit::NetworkStorageManager::getOriginDirectory):
(WebKit::NetworkStorageManager::requestSpace):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp: Renamed from Source/WebKit/NetworkProcess/storage/QuotaManager.cpp.
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::usage):
(WebKit::OriginQuotaManager::requestSpace):
(WebKit::OriginQuotaManager::handleRequests):
(WebKit::OriginQuotaManager::grantWithCurrentQuota):
(WebKit::OriginQuotaManager::usageUpdated):
(WebKit::OriginQuotaManager::grantFastPath):
(WebKit::OriginQuotaManager::didIncreaseQuota):
(WebKit::OriginQuotaManager::resetQuotaUpdatedBasedOnUsageForTesting):
(WebKit::OriginQuotaManager::resetQuotaForTesting):
(WebKit::OriginQuotaManager::reportedQuota const):
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h: Renamed from Source/WebKit/NetworkProcess/storage/QuotaManager.h.
(WebKit::OriginQuotaManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::isActive const):
(WebKit::OriginStorageManager::StorageBucket::hasDataInMemory const):
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::quotaManager):
(WebKit::OriginStorageManager::fileSystemStorageManager):
(WebKit::OriginStorageManager::idbStorageManager):
(WebKit::OriginStorageManager::cacheStorageManager):
(WebKit::OriginStorageManager::backgroundFetchManager):
(WebKit::OriginStorageManager::hasDataInMemory):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/Platform/IPC/WorkQueueMessageReceiver.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration totalQuotaRatio]):
(-[_WKWebsiteDataStoreConfiguration setTotalQuotaRatio:]):
(-[_WKWebsiteDataStoreConfiguration volumeCapacityOverride]):
(-[_WKWebsiteDataStoreConfiguration setVolumeCapacityOverride:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::totalQuotaRatio const):
(WebKit::WebsiteDataStoreConfiguration::setTotalQuotaRatio):
(WebKit::WebsiteDataStoreConfiguration::volumeCapacityOverride const):
(WebKit::WebsiteDataStoreConfiguration::setVolumeCapacityOverride):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/262196@main
